Autopilot/Src/IMU.cpp

```diff
@@ -0,0 +1,205 @@
+      *buffer = REG_GYRO_CONFIG | 0x00;
+      writeData[0] = *buffer;
+      writeData[1] = *setup;
+      write_data(writeData); 
+
+      //set up accelerometer
+      *buffer = REG_ACCEL_CONFIG | 0x00; 
+      writeData[0] = *buffer;
+      write_data(writeData); 
+
+      *setup = 0x08;
+      *buffer = REG_ACCEL_CONFIG_2 | 0x00;
+      writeData[0] = *buffer;
+      writeData[1] = *setup;
+      write_data(writeData); 
+  
+      //Turn on gyroscope and accelerometer
+      *setup = 0x00;
+      *buffer = REG_PWR_MGMT_2 | 0x00;
+      writeData[0] = *buffer;
+      writeData[1] = *setup;
+      write_data(writeData); 
+   }
+
+   accelConversionFactor = ACCEL_SENSITIVITY_4G;
+   gyroConversionFactor = GRYO_SENSITIVITY_500;
+   tempConversionFactor = TEMPERATURE_SENSITIVITY;
+}
+
+void ICM20602::write_data(uint8_t *writeData) { 
+   spi_port->set_slave(0); //Sets CS to HIGH
+   spi_port->exchange_data(writeData, 0x00, 1);
+   spi_port->set_slave(0); //Sets CS to LOW
+}
+
+void ICM20602::get_accel_temp_gyro_reading(float *accx, float *accy, float *accz, float *gyrx, float *gyry, float *gyrz, float *temp) {
+   uint8_t raw_data[14];
+   uint8_t dataFromSensor[2] = {0x00, 0x00};
+
+   //Store both High and Low Byte values
+   int16_t shiftedSensorAccX, shiftedSensorAccY, shiftedSensorAccZ, shiftedSensorTemp, shiftedSensorGyroX, shiftedSensorGyroY, shiftedSensorGyroZ; 
+
+   //Uses the sensor registers to get raw data for all sensors
+   *buffer = REG_ACCEL_XOUT_H | 0x80; 
+   
+   for(int i = 0; i < 14; i++) {
+      dataFromSensor[1] = 0x00; 
+      spi_port->set_slave(0); 
+      sensorSuccess = spi_port->exchange_data(buffer, dataFromSensor, 2); 
+      spi_port->set_slave(0); 
+      *buffer = REG_ACCEL_XOUT_H + 1; //Moves onto next register
+      raw_data[i] = dataFromSensor[1]; //Assigns data byte to the raw_data array for use later
+   }
+   
+   shiftedSensorAccX = (raw_data[0] << 8) + raw_data[1];
+   shiftedSensorAccY = (raw_data[2] << 8) + raw_data[3];
+   shiftedSensorAccZ = (raw_data[4] << 8) + raw_data[5];
+
+   shiftedSensorTemp = (raw_data[6] << 8) + raw_data[7];
+
+   shiftedSensorGyroX = (raw_data[8] << 8) + raw_data[9];
+   shiftedSensorGyroY = (raw_data[10] << 8) + raw_data[11];
+   shiftedSensorGyroZ = (raw_data[12] << 8) + raw_data[13];
+
+   //Converts 16-bit integers to a float. These are the actual measurements
+   *accx = ((float) shiftedSensorAccX)/accelConversionFactor; 
+   *accy = ((float) shiftedSensorAccY)/accelConversionFactor; 
+   *accz = ((float) shiftedSensorAccZ)/accelConversionFactor; 
+
+   *temp = ((float) shiftedSensorTemp)/tempConversionFactor;
+
+   *gyrx = ((float) shiftedSensorGyroX)/gyroConversionFactor;
+   *gyry = ((float) shiftedSensorGyroY)/gyroConversionFactor;
+   *gyrz = ((float) shiftedSensorGyroZ)/gyroConversionFactor;
+}
+
+uint32_t ICM20602::get_current_time() {
+   std::time_t currentTime = std::time(0);
```

Comment:
I talked to Sahil about this, I think we're going to ditch this whole time thing. It's much harder than this to actually get time on a micro-controller and it's of very low value. dw about it right now though, We'll see what Sahil decides on first.